### PR TITLE
use DialOptions, set default timeout, use custom RPC headers

### PR DESCRIPTION
**Pull request by skudasov**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,6 +126,11 @@ In that case you should still pass network name with `-n` flag, especially when
 If `SETH_KEYFILE_PATH` is not set then client will create X ephemeral keys (60 by default, configurable) and won't return any funds.
 Use `SETH_KEYFILE_PATH` for testnets/mainnets and `ephemeral` mode only when testing against simulated network.
 
+In case your RPC node requires custom headers you can set them using `SETH_RPC_HEADERS` in `k=v,k=v` format
+```
+export SETH_RPC_HEADERS=Host=http.com,Host=http.com
```
Comment (lukaszcl) [inline README.md:131]: @skudasov I wonder if you want people to use ~/.testsecrets to load it this? If yes, you need to update `config.ReadFromEnvVar()` in CTF. Another question - do we want custom headers per network?

Comment (lukaszcl) [reply README.md:131]: @skudasov btw, `Host=http.com,Host=http.com` this format does not allow for proper masking on Github CI. We use standard key=value format and mask each value in a `.env` file. Here is the [github step that does it](https://github.com/smartcontractkit/chainlink-github-actions/blob/main/chainlink-testing-framework/run-tests/action.yml#L302)

**Review by lukaszcl — COMMENTED** [review]
@skudasov pls check my comments and reply

File: README.md (hunk 1 above)
Comment (skudasov) [reply README.md:131]: This is infra stuff, no, I want to use our vault or GitHub Secrets directly. People won't know these headers so there is nothing to override.

Comment (lukaszcl) [reply README.md:131]: ok, so if this is not part of Test Secrets then in this case that's fine

**Review by lukaszcl — DISMISSED** [review]

**Review by smickovskid — APPROVED** [review]

File: client.go [hunk 2]
```diff
@@ -133,7 +135,7 @@ func NewClientWithConfig(cfg *Config) (*Client, error) {
 	if len(cfg.Network.URLs) == 0 {
 		return nil, fmt.Errorf("at least one url should be present in config in 'secret_urls = []'")
 	}
-	tr, err := NewTracer(cfg.Network.URLs[0], cs, &abiFinder, cfg, contractAddressToNameMap, addrs)
+	tr, err := NewTracer(cfg.Network.URLs[0], cfg.RPCHeaders, cs, &abiFinder, cfg, contractAddressToNameMap, addrs)
```
Comment (Tofel) [inline client.go:138]: wdyt about just passing cfg here? we could seriously reduce the number of params

File: client.go [hunk 3]
```diff
@@ -240,11 +246,19 @@ func NewClientRaw(
...
+	rpcClient, err := rpc.DialOptions(ctx,
+		cfg.Network.URLs[0],
+		rpc.WithHeaders(cfg.RPCHeaders),
+		rpc.WithHTTPClient(&http.Client{
+			Transport: NewLoggingTransport(),
+		}),
+	)
 	if err != nil {
-		return nil, fmt.Errorf("failed to connect to '%s' due to: %w", cfg.Network.URLs[0], err)
+		return nil, fmt.Errorf("failed to connect RPC client to '%s' due to: %w", cfg.Network.URLs[0], err)
```
Comment (Tofel) [inline client.go:259]: nit: would be nice to have a helper method on `Config` to get `cfg.Network.URLs[0]` just like we have for root address

File: http_logging_transport.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	"time"
+)
+
+// LoggingTransport is a custom transport to log requests and responses
+type LoggingTransport struct {
+	Transport http.RoundTripper
+}
+
+// RoundTrip implements the RoundTripper interface
+func (t *LoggingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (Tofel) [inline http_logging_transport.go:18]: nice!

**Review by Tofel — APPROVED** [review]

**Review by Tofel — APPROVED** [review]

File: client.go (hunk 2 above)
Comment (skudasov) [reply client.go]: done

File: client.go (hunk 3 above)
Comment (skudasov) [reply client.go]: done

**Pull request merged**